File: src/com/dalolorn/sr2modmanager/view/MainController.java
package com.dalolorn.sr2modmanager.view;

import com.dalolorn.sr2modmanager.adapter.RepositoryManager;
import com.dalolorn.sr2modmanager.adapter.Settings;
import javafx.application.Platform;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.VBox;
import javafx.stage.DirectoryChooser;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.stage.StageStyle;

import java.io.*;
import java.util.List;

public class MainController {
	@FXML private TextField urlField;
	@FXML private Label urlLabel;
	@FXML private TextArea modInfo, branchInfo;
	@FXML private ListView<String> branchList;

	public void initialize() {
		branchList.getSelectionModel().selectedItemProperty().addListener((observable, oldValue, newValue) -> setActiveBranch(newValue));

		boolean needsConfig = true;
		try {
			if(!Settings.load()) {
				needsConfig = false;
			}
		} catch(IOException e) {
			Alert msg = new Alert(Alert.AlertType.ERROR, "Could not load config.json! Cause:"  + e.toString());
			e.printStackTrace();
			msg.showAndWait();
		}
		if(needsConfig || !new File(Settings.getInstance().gamePath + File.separator + "Star Ruler 2.exe").exists()) {
			Alert msg = new Alert(Alert.AlertType.WARNING, "Could not autodetect Star Ruler 2 folder! Please navigate to the root folder of your SR2 installation, containing the files 'Star Ruler 2.exe' and 'Star Ruler 2.sh'!");
			msg.showAndWait();
			setSR2Path(new ActionEvent());
		}
	}

	@FXML private void setSR2Path(ActionEvent event) {
		DirectoryChooser chooser = new DirectoryChooser();
		chooser.setInitialDirectory(new File(Settings.getInstance().gamePath));
		chooser.getInitialDirectory().mkdirs();
		chooser.setTitle("Set SR2 Path");
		final File dir = chooser.showDialog(urlField.getScene().getWindow()); // Anything on the window would suffice, urlField was just arbitrarily selected.
		if(dir == null)
			return;
	}

	private void setActiveBranch(String branchName) {
		branchInfo.setText(RepositoryManager.setActiveBranch(branchName));
	}

	@FXML private void connectToRepository(ActionEvent event) {
		// Start preparing the download task.
		final double wndwWidth = 400.0d;
		Label updateLabel = new Label("Loading repository...");
		updateLabel.setPrefWidth(wndwWidth-20);
		updateLabel.setWrapText(true);
		ProgressBar progress = new ProgressBar();
		progress.setPrefWidth(wndwWidth);

		VBox updatePane = new VBox();
		updatePane.setPadding(new Insets(10, 10, 30, 10));
		updatePane.setSpacing(5.0d);
		updatePane.getChildren().addAll(updateLabel, progress);

		Stage taskUpdateStage = new Stage(StageStyle.UTILITY);
		taskUpdateStage.initModality(Modality.APPLICATION_MODAL);
		taskUpdateStage.setScene(new Scene(updatePane));
		taskUpdateStage.show();

		Task updateTask = new Task<Void>() {
			@Override
			protected Void call() {
				String repoURL;
				RepositoryManager.TextHandler errorHandler = error -> Platform.runLater(() -> {
					Alert msg = new Alert(Alert.AlertType.ERROR, error);
					msg.show();
				});

				if (
					(repoURL = RepositoryManager.connectToRepository
						(
							urlField.getText(),
							this::updateMessage,
							errorHandler
						)
					) == null
				) return null;

				updateMessage("Fetching branch list...");
				if(!getBranches(errorHandler)) return null;

				updateMessage("Getting mod description...");
				if(!getDescription(errorHandler)) return null;

				Platform.runLater(() -> urlLabel.setText("Connected to " + repoURL));
				return null;
			}

			@Override
			protected void succeeded() {
				taskUpdateStage.close();
			}
		};

		progress.progressProperty().bind(updateTask.progressProperty());
		updateLabel.textProperty().bind(updateTask.messageProperty());

		taskUpdateStage.show();
		new Thread(updateTask).start();
	}

	@FXML private void installMod(ActionEvent event) {
		// Start preparing the install task.
		final double wndwWidth = 400.0d;
		Label installLabel = new Label("Preparing to install mod...");
		installLabel.setPrefWidth(wndwWidth-20);
		installLabel.setWrapText(true);
		ProgressBar progress = new ProgressBar();
		progress.setPrefWidth(wndwWidth);

		VBox installPane = new VBox();
		installPane.setPadding(new Insets(10, 10, 30, 10));
		installPane.setSpacing(5.0d);
		installPane.getChildren().addAll(installLabel, progress);

		Stage taskInstallStage = new Stage(StageStyle.UTILITY);
		taskInstallStage.initModality(Modality.APPLICATION_MODAL);
		taskInstallStage.setScene(new Scene(installPane));
		taskInstallStage.show();
		Task installTask = new Task<Void>() {
			@Override
			protected Void call() {
				updateMessage("Preparing to install mod...");
				RepositoryManager.installMod(
						warning -> Platform.runLater(() -> {
							Alert msg = new Alert(Alert.AlertType.WARNING, warning);
							msg.show();
						}),
						this::updateMessage,
						info -> Platform.runLater(() -> {
							Alert msg = new Alert(Alert.AlertType.INFORMATION, info);
							msg.show();
						}),
						error -> Platform.runLater(() -> {
							Alert msg = new Alert(Alert.AlertType.ERROR, error);
							msg.show();
						})
				);
				return null;
			}

			@Override
			protected void succeeded() {
				taskInstallStage.close();
			}
		};

		progress.progressProperty().bind(installTask.progressProperty());
		installLabel.textProperty().bind(installTask.messageProperty());

		taskInstallStage.show();
		new Thread(installTask).start();
	}

	private boolean getBranches(RepositoryManager.TextHandler errorHandler) {
		List<String> tagNames = RepositoryManager.getBranches(errorHandler);
		if(tagNames == null) return false;

		Platform.runLater(() -> {
			branchList.getItems().clear();
			branchList.getItems().addAll(tagNames);
		});
		return true;
	}

	private boolean getDescription(RepositoryManager.TextHandler errorHandler) {
		String description = RepositoryManager.getDescription(errorHandler);
		if(description == null) return false;

		Platform.runLater(() -> {
			modInfo.setText(description);

			branchInfo.setText("No branches selected");
		});
		return true;
	}

	@FXML private void openRepository(ActionEvent actionEvent) {
		// Get repository location.
		DirectoryChooser chooser = new DirectoryChooser();
		chooser.setInitialDirectory(new File("repositories/"));
		chooser.getInitialDirectory().mkdirs();
		chooser.setTitle("Choose a Repository");
		final File dir = chooser.showDialog(urlField.getScene().getWindow()); // Anything on the window would suffice, urlField was just arbitrarily selected.
		if(dir == null)
			return;

		// Start preparing the download task.
		final double wndwWidth = 400.0d;
		Label updateLabel = new Label("Loading repository...");
		updateLabel.setPrefWidth(wndwWidth-20);
		updateLabel.setWrapText(true);
		ProgressBar progress = new ProgressBar();
		progress.setPrefWidth(wndwWidth);

		VBox updatePane = new VBox();
		updatePane.setPadding(new Insets(10, 10, 30, 10));
		updatePane.setSpacing(5.0d);
		updatePane.getChildren().addAll(updateLabel, progress);

		Stage taskUpdateStage = new Stage(StageStyle.UTILITY);
		taskUpdateStage.initModality(Modality.APPLICATION_MODAL);
		taskUpdateStage.setScene(new Scene(updatePane));
		taskUpdateStage.show();

		Task updateTask = new Task<Void>() {
			@Override
			protected Void call() {
				String repoURL;
				RepositoryManager.TextHandler errorHandler = error -> Platform.runLater(() -> {
					Alert msg = new Alert(Alert.AlertType.ERROR, error);
					msg.show();
				});

				if (
						(repoURL = RepositoryManager.openRepository
								(
										dir,
										errorHandler
								)
						) == null
				) return null;

				updateMessage("Fetching branch list...");
				if(!getBranches(errorHandler)) return null;

				updateMessage("Getting mod description...");
				if(!getDescription(errorHandler)) return null;

				Platform.runLater(() -> urlLabel.setText("Connected to " + repoURL));
				return null;
			}

			@Override
			protected void succeeded() {
				taskUpdateStage.close();
			}
		};

		progress.progressProperty().bind(updateTask.progressProperty());
		updateLabel.textProperty().bind(updateTask.messageProperty());

		taskUpdateStage.show();
		new Thread(updateTask).start();
	}

	@FXML private void close(ActionEvent actionEvent) {
		Alert dialog = new Alert(Alert.AlertType.CONFIRMATION, "Quit SR2 Mod Manager?");
		ObservableList<ButtonType> dlgButtons = dialog.getDialogPane().getButtonTypes();
		dlgButtons.clear();
		dlgButtons.add(ButtonType.YES);
		dlgButtons.add(ButtonType.NO);
		dialog.setHeaderText("Quit SR2MM");

		dialog.showAndWait()
				.filter(response -> response == ButtonType.YES)
				.ifPresent(response -> System.exit(0));
	}

	@FXML private void deleteRepository(ActionEvent actionEvent) {
		Alert dialog = new Alert(Alert.AlertType.CONFIRMATION, "Are you sure you want to delete this repository from your computer? You will not be able to access it again until you redownload it.");
		ObservableList<ButtonType> dlgButtons = dialog.getDialogPane().getButtonTypes();
		dlgButtons.clear();
		dlgButtons.add(ButtonType.YES);
		dlgButtons.add(ButtonType.NO);
		dialog.setHeaderText("Delete Repository?");

		dialog.showAndWait()
				.filter(response -> response == ButtonType.YES)
				.ifPresent(response -> {
						if(RepositoryManager.deleteRepository(error -> {
							Alert msg = new Alert(Alert.AlertType.ERROR, error);
							msg.show();
						})) {
							branchList.getItems().clear();
							modInfo.setText("");
							branchInfo.setText("");
						}
				});
	}

	@FXML private void about(ActionEvent actionEvent) {
		Alert msg = new Alert(Alert.AlertType.INFORMATION,
				"SR2 Mod Manager by Dalo Lorn\n" +
						"\n" +
						"Version: 1.1.1\n\n" +
						"GitHub: https://github.com/DaloLorn/SR2ModManager\n" +
						"Discord: https://discord.gg/sUJKJDc\n" +
						"Patreon: https://patreon.com/rising_stars_sr2");
		msg.setHeaderText("About SR2 Mod Manager");
		msg.show();
	}
}

File: src/main/java/com/dalolorn/sr2modmanager/view/MainController.kt
package com.dalolorn.sr2modmanager.view

import com.dalolorn.sr2modmanager.adapter.ModInstaller
import com.dalolorn.sr2modmanager.adapter.ModInstaller.TextHandler
import com.dalolorn.sr2modmanager.adapter.Settings
import com.dalolorn.sr2modmanager.adapter.Utils
import javafx.application.Platform
import javafx.concurrent.Task
import javafx.fxml.FXML
import javafx.geometry.Insets
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.control.Alert.AlertType
import javafx.scene.layout.VBox
import javafx.stage.DirectoryChooser
import javafx.stage.Modality
import javafx.stage.Stage
import javafx.stage.StageStyle
import java.io.File
import java.io.IOException
import kotlin.system.exitProcess

class MainController {
    // Shared UI components.
    @FXML private val playButton: Button? = null

    @FXML private val tabs: TabPane? = null
    @FXML private val modsTab: Tab? = null
    @FXML private val installerTab: Tab? = null
    @FXML private val osrTab: Tab? = null

    @FXML private val openRepositoryItem: MenuItem? = null
    @FXML private val deleteRepositoryItem: MenuItem? = null
    @FXML private val uninstallModItem: MenuItem? = null

    @FXML private val modsPane: ModManagerPane? = null
    @FXML private val installerPane: ModInstallerPane? = null
    @FXML private val osrPane: OpenSRManagerPane? = null

    fun initialize() {
        tabs!!.selectionModel.selectedItemProperty().addListener { _, _, newTab ->
            deleteRepositoryItem!!.isDisable = newTab !== installerTab || !ModInstaller.hasRepo()
            uninstallModItem!!.isDisable = newTab !== installerTab
        }
        osrPane!!.playButton = playButton

        var needsConfig = true
        try {
            needsConfig = !Settings.load()
        } catch (e: IOException) {
            val msg: Alert = ResizableAlert(AlertType.ERROR, "Could not load config.json! Cause: $e")
            e.printStackTrace()
            msg.showAndWait()
        }

        if (
            needsConfig
            || !File(Settings.instance.gamePath, "Star Ruler 2.exe").exists()
            && !File(Settings.instance.gamePath, "StarRuler2.sh").exists()
        ) {
            val msg = ResizableAlert(
                AlertType.WARNING,
                "Could not detect Star Ruler 2 launchers! Please navigate to the root folder of your SR2 installation, containing the files 'Star Ruler 2.exe' and/or 'StarRuler2.sh'!"
            )
            msg.showAndWait()
            setSR2Path()
        }
    }

    @FXML
    private fun setSR2Path() {
        osrPane!!.setSR2Path()
    }

    @FXML
    private fun openRepository() {
        // Get repository location.
        val chooser = DirectoryChooser()
        chooser.initialDirectory = File("repositories/")
        chooser.initialDirectory.mkdirs()
        chooser.title = "Choose a Repository"
        val dir = chooser.showDialog(playButton!!.scene.window)
            ?: return // Anything on the window would suffice, playButton was just arbitrarily selected.

        executeTask("Loading repository...") { taskUpdateStage: Stage ->
            object : Task<Unit>() {
                override fun call() {
                    var repoURL: String
                    val errorHandler = TextHandler { text ->
                        Platform.runLater {
                            val msg: Alert = ResizableAlert(AlertType.ERROR, text)
                            msg.show()
                        }
                    }
                    if (ModInstaller.openRepository(
                            dir,
                            errorHandler
                        ).also { repoURL = it!! } == null
                    ) return
                    updateMessage("Fetching branch list...")
                    if (!installerPane!!.getBranches(errorHandler)) return
                    updateMessage("Getting mod description...")
                    if (!installerPane.getDescription(errorHandler)) return
                    Platform.runLater { installerPane.setUrlText("Connected to $repoURL") }
                    return
                }

                override fun succeeded() {
                    taskUpdateStage.close()
                }
            }
        }
    }

    @FXML
    private fun close() {
        val dialog = ResizableAlert(AlertType.CONFIRMATION, "Quit OpenSR Launchpad?")
        val dlgButtons = dialog.dialogPane.buttonTypes
        dlgButtons.clear()
        dlgButtons.add(ButtonType.YES)
        dlgButtons.add(ButtonType.NO)
        dialog.headerText = "Quit OpenSR Launchpad"
        dialog.showAndWait()
            .filter { response: ButtonType -> response == ButtonType.YES }
            .ifPresent { exitProcess(0) }
    }

    @FXML
    private fun deleteRepository() {
        if (installerTab!!.isSelected) {
            installerPane!!.deleteRepository()
        }
    }

    @FXML
    private fun about() {
        val msg: Alert = ResizableAlert(
            AlertType.INFORMATION,
            """
                  OpenSR Launchpad
                  
                  Version: 2.0.0
                  
                  GitHub: https://github.com/OpenSRProject/OpenStarRuler-Launchpad
                  Discord: https://discord.gg/sUJKJDc
                  """.trimIndent()
        )
        msg.headerText = "About OpenSR Launchpad"
        msg.show()
    }

    @FXML
    private fun startGame() {
        val launcher =
            File(Settings.instance.gamePath, if (Utils.IS_WINDOWS) "Star Ruler 2.exe" else "StarRuler2.sh")
        if (!launcher.exists()) ResizableAlert(
            AlertType.ERROR,
            "Cannot start Star Ruler 2: Couldn't find platform-appropriate loader!"
        )
            .showAndWait()
        if (!launcher.canExecute()) ResizableAlert(
            AlertType.ERROR,
            "Cannot start Star Ruler 2: Couldn't run game loader!"
        )
            .showAndWait()
        try {
            ProcessBuilder(launcher.absolutePath)
                .directory(launcher.parentFile)
                .start()
        } catch (e: Exception) {
            val msg = ResizableAlert(AlertType.ERROR, String.format("Cannot start Star Ruler 2: %s", e))
            e.printStackTrace()
            msg.showAndWait()
        }
    }

    @FXML
    fun uninstallMod() {
        if (installerTab!!.isSelected) {
            installerPane!!.uninstallMod()
        }
    }

    companion object {
        fun executeTask(
            initialText: String,
            taskBuilder: (Stage) -> Task<Unit>
        ) {
            // Start preparing the task.
            val wndwWidth = 400.0
            val taskLabel = Label(initialText)
            taskLabel.prefWidth = wndwWidth - 20
            taskLabel.isWrapText = true

            val progress = ProgressBar()
            progress.prefWidth = wndwWidth

            val taskPane = VBox()
            taskPane.padding = Insets(10.0, 10.0, 30.0, 10.0)
            taskPane.spacing = 5.0
            taskPane.children.addAll(taskLabel, progress)

            val taskStage = Stage(StageStyle.UTILITY)
            taskStage.initModality(Modality.APPLICATION_MODAL)
            taskStage.scene = Scene(taskPane)
            taskStage.show()

            val task = taskBuilder(taskStage)
            progress.progressProperty().bind(task.progressProperty())
            taskLabel.textProperty().bind(task.messageProperty())
            taskStage.show()
            Thread(task).start()
        }
    }
}